src/audio/mux_generic.c

```diff
@@ -0,0 +1,273 @@
+					if (data[j].mask[out_ch] & BIT(in_ch)) {
+						src = buffer_read_frag_s16(source, i * data[j].num_channels + in_ch);
+						sample += *src;
+					}
+				}
+			}
+			dst = buffer_write_frag_s16(sink,
+				i * data->num_channels + out_ch);
+			*dst = sat_int16(sample);
+		}
+	}
+}
+
+static void demux_s24le(struct comp_dev *dev, struct comp_buffer *sink,
+			struct comp_buffer **sources, uint32_t frames,
+			struct mux_stream_data *data)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	uint8_t i;
+	uint8_t j;
+	uint8_t in_ch;
+	uint8_t out_ch;
+	int32_t *src;
+	int32_t *dst;
+	int32_t sample;
+
+	for (i = 0; i < frames; i++) {
+		for (out_ch = 0; out_ch < cd->num_channels; out_ch++) {
+			sample = 0;
+			for (j = 0; j < MUX_MAX_STREAMS; j++) {
+				source = sources[j];
+				if (!source)
+					continue;
+
+				for (in_ch = 0; in_ch < data[j].num_channels; in_ch++) {
+					if (data[j].mask[out_ch] & BIT(in_ch)) {
+						src = buffer_read_frag_s32(source, i * data[j].num_channels + in_ch);
+						sample += sign_extend_s24(*src);
+					}
+				}
+			}
+			dst = buffer_write_frag_s32(sink,
+				i * data->num_channels + out_ch);
+			*dst = sat_int24(sample);
+		}
+	}
+}
+
+static void demux_s32le(struct comp_dev *dev, struct comp_buffer *sink,
+			struct comp_buffer **sources, uint32_t frames,
+			struct mux_stream_data *data)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	uint8_t i;
+	uint8_t j;
+	uint8_t in_ch;
+	uint8_t out_ch;
+	int32_t *src;
+	int32_t *dst;
+	int64_t sample;
+
+	for (i = 0; i < frames; i++) {
+		for (out_ch = 0; out_ch < cd->num_channels; out_ch++) {
+			sample = 0;
+			for (j = 0; j < MUX_MAX_STREAMS; j++) {
+				source = sources[j];
+				if (!source)
+					continue;
+
+				for (in_ch = 0; in_ch < data[j].num_channels; in_ch++) {
+					if (data[j].mask[out_ch] & BIT(in_ch)) {
+						src = buffer_read_frag_s32(source, i * data[j].num_channels + in_ch);
+						sample += *src;
+					}
+				}
+			}
+			dst = buffer_write_frag_s32(sink,
+				i * data->num_channels + out_ch);
+			*dst = sat_int32(sample);
+		}
+	}
+}
+
+const struct comp_func_map mux_func_map[] = {
+	{ SOF_IPC_FRAME_S16_LE, &mux_s16le, &demux_s16le },
+	{ SOF_IPC_FRAME_S24_4LE, &mux_s24le, &demux_s24le },
+	{ SOF_IPC_FRAME_S32_LE, &mux_s32le, &demux_s32le },
+};
+
+mux_func mux_get_processing_function(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint8_t i;
+
+	for (i = 0; i < ARRAY_SIZE(mux_func_map); i++) {
+		if (cd->frame_format == mux_func_map[i].frame_format)
+			return mux_func_map[i].mux_proc_func;
+	}
+
+	return NULL;
+}
+
+demux_func demux_get_processing_function(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint8_t i;
+
+	for (i = 0; i < ARRAY_SIZE(mux_func_map); i++) {
+		if (cd->frame_format == mux_func_map[i].frame_format)
+			return mux_func_map[i].demux_proc_func;
+	}
+
+	return NULL;
+}
```

Comment:
if we only have one function in the comp_data, we can pass the mode, and pick the appropriate function instead of 2 separate functions which have the exact same logic